stellargraph/link_prediction.py

```diff
@@ -0,0 +1,185 @@
+            self._classifiers[binary_operator] = clf
+
+    def predict(self, link_examples, transform_node=None, binary_operators=None):
+        """
+        Predict links using trained classifiers.
+
+        Args:
+            link_examples (list of tuple): Tuples of links to predict (source, target)
+            transform_node (callable, optional): If provided, use this function to transform node
+                IDs into node features instead of the one used to train the classifiers.
+            binary_operators (list of str, optional): If provided, only predict using these binary
+                operators regardless of which classifiers have been trained. By default, all trained
+                classifiers are used.
+
+        Returns:
+            Dict of predictions grouped by binary operator
+        """
+        node_feature_pairs = self._transform_node_pairs(link_examples, transform_node)
+        predictions = dict()
+
+        if binary_operators is None:
+            binary_operators = self._classifiers.keys()
+
+        for binary_operator in binary_operators:
+            if binary_operator not in self._classifiers:
+                warnings.warn(
+                    f"Skipping binary operator '{binary_operator}': "
+                    f"no trained classifier available for this operator."
+                )
+            else:
+                classifier = self._classifiers[binary_operator]
+                operator_func = _get_operator(binary_operator)
+                link_features = [
+                    operator_func(src, dst) for src, dst in node_feature_pairs
+                ]
+                predictions[binary_operator] = classifier.predict_proba(link_features)
+
+        return predictions
+
+    def evaluate_roc_auc(
+        self, link_examples, link_labels, transform_node=None, binary_operators=None
+    ):
+        """
+        Evaluate trained classifiers using the provided test set.
+
+        Args:
+            link_examples (list of tuple): Tuples of test set links to predict (source, target)
+            link_labels (list of int): List of labels corresponding to link examples - 1 for positive,
+                0 for negative.
+            transform_node (callable, optional): If provided, use this function to transform node
+                IDs into node features instead of the one used to train the classifiers.
+            binary_operators (list of str, optional): If provided, only predict using these binary
+                operators regardless of which classifiers have been trained. By default, all trained
+                classifiers are used.
+
+        Returns:
+            Dict of Scores grouped by binary operator
+        """
+        predictions = self.predict(link_examples, transform_node, binary_operators)
+        scores = dict()
+        for binary_operator, predicted in predictions.items():
+            if self._classifiers[binary_operator].classes_[0] == 1:
```

Comment:
This `if` statement seems a bit peculiar, what's special about the first class having value `1`?

Reply:
You need to know which column indicates the predictions for the edge label, i.e., the label you are actually predicting, indicated by the label 1. Then you can pass the correct `y_score` data to the `roc_auc_score` method. One could train the classifier to predict no-edges instead but that would be indicated by label 0 and calculating the AUC ROC score would require picking  the other column.

Reply:
Ah, right, so this `if` block is actually a manual version of:

```python
positive_column = self._classifiers[binary_operator].classes_.index(1)
scores[binary_operator] = roc_auc_score(link_labels, predicted[:, positive_column])
```